Preserve stdout byte stream for native commands

SeeminglyScience opened the pull request:
With this change, any native commands in a pipeline will be able to detect if the downstream command is also native. In that case (and when stderr is not merged into stdout), the native command will write directly to the input stream of the downstream command.

Also added in this change is the ability to pipe bytes directly to a native command, e.g.

```powershell
0xFFuy | native
[byte[]](0..255) | native

# Passing as a strongly typed byte array with unary comma also works for performance
,[byte[]](0..255) | native
```

<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->

## PR Context

Left to do:

- [x] Hook up provider path resolution and error handling
- [x] See if it's feasible to wrap as an experimental feature
- [x] Fix handling for manually created pipelines (e.g. `PowerShell.AddCommand`) 
- [x] Investigate `Stream.CopyToAsync` (thank you @daxian-dbw!)
- [x] Fix steppable pipelines
- [x] Tests
- [x] Clean up

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): PSNativeCommandPreserveBytePipe
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: MicrosoftDocs/PowerShell-Docs#10134
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

SeeminglyScience committed:
Preserve stdout byte stream for `native | native`

Also preserve for `native > file.txt`.

SeeminglyScience commented on src/System.Management.Automation/engine/AsyncByteStreamDrainer.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace System.Management.Automation;
+
+internal sealed class AsyncByteStreamDrainer : IDisposable
```
Really don't like this name. `StreamReader` implies string or I'd use that. Open to suggestions

AE1020 replied:
AsyncByteStreamRepeater?

SeeminglyScience replied:
`AsyncByteStreamTransfer` maybe?

SeeminglyScience commented on src/System.Management.Automation/engine/BytePipe.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace System.Management.Automation;
+
+internal interface IStreamSource
```
This whole file is a little over engineered and needs to be rethought a bit

SeeminglyScience commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -403,8 +411,15 @@ internal NativeCommandParameterBinderController NativeParameterBinderController
         /// </summary>
         internal override void Prepare(IDictionary psDefaultParameterValues)
         {
+            if (_isPreparedCalled)
+            {
+                return;
+            }
+
             _isPreparedCalled = true;
 
+            DownStreamNativeCommand?.Prepare(psDefaultParameterValues);
```
This can probably be removed. Initially I wasn't lazily retrieving the stdout stream so the downstream native command had to be prepared before the upstream. I believe I've fixed that

SeeminglyScience commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1714,7 +1768,18 @@ internal class ProcessOutputHandler
+        {
+        }
+
+        public ProcessOutputHandler(
+            Process process,
+            BlockingCollection<ProcessOutputObject> queue,
+            BytePipe stdOutDestination,
+            out AsyncByteStreamDrainer stdOutDrainer)
```
I'm trying not to *too* drastically alter the classes here but `ProcessOutputHandler` is a bit non-typical already. We were just creating an instance of it and not saving it anywhere so the constructor sort of functioned like a static method. I need to take a closer look at if it's worth rethinking this class in general as this out param doesn't help things.

SeeminglyScience commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -207,9 +214,20 @@ internal static class PipelineOps
             bool redirectedInformation = false;
             if (redirections != null)
             {
+                foreach (var redirection in redirections) {
+                    if (redirection is MergingRedirection)
+                    {
+                        redirection.Bind(pipe, commandProcessor, context);
+                    }
+                }
```
The double enumeration here was a quick band-aid for `native 2>&1 > file.txt`. In that scenario we want to revert to the current string reading behavior so I needed to process merging redirections before I tried to create a specialized byte pipe for a file stream.

I need to find implement a better pattern here.

SeeminglyScience commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -723,7 +757,7 @@ internal static SteppablePipeline GetSteppablePipeline(PipelineAst pipelineAst,
             foreach (var commandTuple in commandTuples)
             {
-                var commandProcessor = AddCommand(pipelineProcessor, commandTuple.Item2.ToArray(), commandTuple.Item1, commandTuple.Item3.ToArray(), context);
+                var commandProcessor = AddCommand(pipelineProcessor, commandTuple.Item2.ToArray(), commandTuple.Item1, commandTuple.Item3.ToArray(), context, false);
```
Haven't touched steppable pipelines yet, that likely doesn't work atm.

SeeminglyScience commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -1047,6 +1081,15 @@ public override string ToString()
+                && FromStream is RedirectionStream.Output
+                && !string.IsNullOrWhiteSpace(File))
+            {
+                nativeCommand.StdOutDestination = FileBytePipe.Create(File, Appending);
```
This is currently just opening a file stream. It needs path resolution to be done through `FileSystemProvider` and it needs to surface errors similarly to how `Out-File` does today.

SeeminglyScience commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1981,14 +2058,31 @@ internal void Add(object input)
+            if (_inputFormat is not NativeCommandIOFormat.Text)
             {
-                AddTextInput(input);
+                AddXmlInput(input);
+                return;
             }
-            else // Xml
+
+            object baseObjInput = input;
+            if (input is PSObject pso)
             {
-                AddXmlInput(input);
+                baseObjInput = pso.BaseObject;
             }
+
+            if (baseObjInput is byte[] bytes)
+            {
+                _streamWriter.BaseStream.Write(bytes, 0, bytes.Length);
+                return;
+            }
+
+            if (baseObjInput is byte b)
+            {
+                _streamWriter.BaseStream.WriteByte(b);
+                return;
+            }
+
+            AddTextInput(input);
```
This bit enables piping bytes directly to native commands. Might need to be in a different PR, not strictly in scope.

SeeminglyScience commented on src/System.Management.Automation/engine/BytePipe.cs:
```diff
@@ -0,0 +1,100 @@
+
+    internal static FileBytePipe Create(string fileName, bool append)
+    {
+        return new FileBytePipe(
+            new FileStream(
+                fileName,
+                append ? FileMode.Append : FileMode.OpenOrCreate,
+                FileAccess.Write,
+                FileShare.ReadWrite | FileShare.Delete));
```
The options here likely need to be changed to match `Out-File`

SeeminglyScience commented on src/System.Management.Automation/engine/BytePipe.cs:
```diff
@@ -0,0 +1,100 @@
+                sourceStream,
+                (bytes, _) =>
+                {
+                    lock (_syncObject)
+                    {
+                        GetStream().Write(bytes);
+                    }
```
The lock might not be needed here. Originally this was due to the drainers for `stdout` and `stderr` writing to the same stream when redirected, but now we revert to string reading for that.

AE1020 commented:
Glad to see this being taken seriously!  It is important for generic shells--especially ones that aspire to work cross platform--to be able to pass data through agnostically.

*(I cloned and built Powershell just to try this PR! 😃)*

The stdout redirection did appear to work on my tests (which require not introducing CR into LF-only streams).  That is good!  I was able to pipe | and redirect > successfully.

However, the powershell I built seems to put the redirected native outputs wherever powershell was started up, not in the current directory.  So if run from some directory like:

    C:\projects\> C:\projects\powershell\src\powershell-win-core\bin\Debug\net7.0\win7-x64\publish\pwsh
    PowerShell 7.3.0-preview.3-267-g4faa5f99e3f6c7f959da35ce264959bb43017126
    PS C:\Projects> cd test
    PS C:\Projects\test> git --version > gitver.txt
    PS C:\Projects\test> cat gitver.txt
    Get-Content: Cannot find path 'C:\Projects\test\gitver.txt' because it does not exist.

But we find the file in the original directory:

    PS C:\Projects\test> cat ..\gitver.txt
    git version 2.36.1.windows.1

Piping commands that aren't native output, e.g. **`echo "hello" > echotest.txt`** act as expected, and go to the current directory.

Redirecting 2> for errors is introducing CRs when CMD.EXE does not give them, and does not have the directory issue...so I assume it is not running the new handling.  (As per my post on the discussion thread, my vote is very much that stderr be covered by the same handling!)

SeeminglyScience commented:
> Glad to see this being taken seriously! It is important for generic shells--especially ones that aspire to work cross platform--to be able to pass data through agnostically.
>
> _(I cloned and built Powershell just to try this PR! 😃)_

❤️ 

> However, the powershell I built seems to put the redirected native outputs wherever powershell was started up, not in the current directory. So if run from some directory like:

Ah yeah that makes sense. That'd be part of what gets fixed with my comment above about hooking up path resolution via `FileSystemProvider`.

> Redirecting 2> for errors is introducing CRs when CMD.EXE does not give them, and does not have the directory issue...so I assume it is not running the new handling. (As per my post on the discussion thread, my vote is very much that stderr be covered by the same handling!)

Yeah I can definitely understand the desire, but the only real way to do that with consistency would be to rewrite the `Process` class from scratch with p/invoke code specific to every platform. While that does genuinely sound like a fun thing to write, that's a ton of extra code to maintain for the level of impact it would have. I asked in that thread because if there were show stoppers around reverting to current behavior for stderr, I'd likely have to scrap the idea entirely.

Maybe in the future dotnet will add something to the `SD.Process` class to make it feasible, but for now this is an acceptable compromise imo.

AE1020 commented:
> the only real way to do that with consistency would be to rewrite the Process class from scratch with p/invoke code specific to every platform

Can you explain in more detail why this is significantly different from the stdout case?

SeeminglyScience commented:
> Can you explain in more detail why this is significantly different from the stdout case?

The biggest technical issue is around merging stderr into stdout. So the ideal way this would work (using Windows APIs as an example) is I would create a single pipe (or file for redirection) with `kernel32!CreatePipe` (or `kernel32!CreateFile`) that I could pass to `kernel32!CreateProcess` for both stdout and stderr. I don't have a generic way to do this without writing a lot of platform specific code so instead I'm reading from the `Stream` created by `System.Diagnostics.Process` for `stdout`.

Doing the same for `stderr` and trying to flush them both into the downstream's stdin `Stream` leads to a significant amount of inconsistency in order. That inconsistency already exists in the current string reading behavior but it's made worse when dealing with bytes directly as you can end up writing half of a line (or even half of a unicode glyph) from stdout and then another half from stderr.

This doesn't directly exclude lighting up the `native 2> log.txt` scenario and reverting only for `native 2>&1`, but it does make it less clear cut. It's easy to understand that "error is text, so it's treated like strings, and when you merge it makes stdout also like strings". It's harder to understand "neither are treated like strings unless you merge them together" if that makes sense.

It's also not necessarily a simple switch I'd be flipping to enable it for stderr as well. There are a lot of places it needs to be special cased in pipeline creation similarly to how I'm doing it for stdout but subtly different. If/when this PR is merged I'll create a separate issue that is specifically for preserving bytes when redirecting stderr so the engine WG can discuss it. This PR will be for stdout specifically, but that doesn't necessarily mean another PR can't address it.

AE1020 commented:
The stdout case probably would satisfy most people, whose concerns are more about binary preservation with things like curl/zip.  My issue with cross-platform consistency on CRLF is its own crusade, which is distinct from this if one is going to actually "use" PowerShell in a bigger sense.  But it happens to be measurably better with this change.

I can see that interleaving the streams raises a lot of issues.  I'm not clear on how it *ever* works (are UNIX shells at risk of doing half-codepoints in UTF-8?  If not, how?  Would their guarantee be accomplished by line buffering vs. understanding UTF-8 encodings specifically?  If so, what happens when the line buffer size is exceeded?)  😦  Sounds like something one would find a lot of inconvenient truths by reading up on...

SeeminglyScience commented:
> I can see that interleaving the streams raises a lot of issues. I'm not clear on how it _ever_ works (are UNIX shells at risk of doing half-codepoints in UTF-8? If not, how? Would their guarantee be accomplished by line buffering vs. understanding UTF-8 encodings specifically? If so, what happens when the line buffer size is exceeded?) 😦 Sounds like something one would find a lot of inconvenient truths by reading up on...

The trouble mostly comes from utilizing `System.Diagnostics.Process`. When you redirect a stream using that API, it creates a pipe for each stream you've redirected. Then I would be asynchronously reading from both of these separate streams and trying to merge the output after the fact. These streams will also have different buffer sizes and flush intervals.

It's much easier when you are synchronously reading from and writing to a single pipe, as most platform specific shells with less output processing will do.

AE1020 commented:
@adityapatwardhan What sort of feedback would be needed for PowerShell to prioritize reviewing this?

It is a rather popular request (see #1908, and touched upon by some other issues), where these seem like things people expect to work in a shell:

    curl.exe http://whatever/a.png > a.png

    node a.js | gzip -c > out.gz

Modulo the output directory I found when trying it, it seemed to work for that.

SeeminglyScience commented:
> What sort of feedback would be needed for PowerShell to prioritize reviewing this?

It's not finished yet, it's a work in progress PR. You can ignore the bot, if it closed it then I'd just reopen

SeeminglyScience committed:
Add filesystem provider path resolution

SeeminglyScience commented on src/System.Management.Automation/engine/BytePipe.cs:
```diff
@@ -0,0 +1,124 @@
+                out _,
+                true);
+        }
+        catch (Exception e) when (e.Data.Contains(typeof(ErrorRecord)))
+        {
+            // The error record is attached to the exception when thrown to preserve
+            // the call stack.
+            ErrorRecord? errorRecord = e.Data[typeof(ErrorRecord)] as ErrorRecord;
+            if (errorRecord is null)
+            {
+                throw;
+            }
+
+            e.Data.Remove(typeof(ErrorRecord));
+            throw new RuntimeException(null, e, errorRecord);
+        }
```
This was the cleanest way I could think of to throw the underlying error record/exception while retaining the same exception type as `Out-File` and the stack trace. Open to other ideas

iSazonov commented:
> Combine sequential native pipline elements and treat them as one pipeline element. Link them with IO redirection directly instead of powershell pipeline.

This was in https://github.com/PowerShell/PowerShell/issues/559#issuecomment-240957563
If I understand write it is that we want, but can not implement due to Process class limitations, so we would have to fall in low APIs?

SeeminglyScience commented:
> This was in [#559 (comment)](https://github.com/PowerShell/PowerShell/issues/559#issuecomment-240957563) If I understand write it is that we want, but can not implement due to Process class limitations, so we would have to fall in low APIs?

So it kind of depends how you interpret what @vors was saying there. In a sense, that's what I'm doing in this PR.

In another sense (which is what you're getting at) there is another route that would be a much bigger lift but would be closer to how other shells implement it.

On Windows that would be using `CreatePipe` to create a single pipe that we pass to `CreateProcess` for both the upstream's stdout and the downstream's stdin. And something somewhat similar on *nix with `pipe`.

That would require rewriting a significant chunk of the code as we wouldn't be able to use `System.Diagnostics.Process` at all really. I'm definitely interested to see if I can make that work, but not in this PR. Probably something I need to pursue on my own time, and tbh highly likely to be too great of an additional maintenance burden for the value add anyway.

SeeminglyScience committed:
Fix path resolution

SeeminglyScience committed:
Added experimental feature

SeeminglyScience committed:
Fix pipeline handling outside of the compiler

SeeminglyScience commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -160,7 +161,13 @@ internal static class PipelineOps
+            {
+                isNativeCommand = true;
+                nativeCommandProcessor.UpstreamIsNativeCommand = lastCommandWasNative;
+            }
```
> I guess this won't work for pipelines invoked via `PowerShell.AddCommand`, given the pipeline construction won't go through the compiler code.

Good catch! Fixed with [`a6f0b1c` (#17857)](https://github.com/PowerShell/PowerShell/pull/17857/commits/a6f0b1c51695813bb5297f6f3e1e4d2f6dcb351c) ❤️

SeeminglyScience committed:
Merge branch 'master' into preserve-native-stdout

SeeminglyScience committed:
Fix merging redirection fallback behavior

Merging stderr into stdout was resulting in a dead lock due to the order
that `NativeCommandProcessor`'s were initializing. This was fixed by no
longer requiring upstream commands to be `Prepare`ed before downstream
commands.

SeeminglyScience committed:
Add byte read/write to TestExe and tests

SeeminglyScience commented on test/powershell/engine/Basic/NativeCommandBytePiping.Tests.ps1:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    It 'Bytes are retained when using SteppablePipeline' {
```
This test currently fails because I haven't fixed steppable pipeline yet

SeeminglyScience commented on test/tools/TestExe/TestExe.cs:
```diff
@@ -41,6 +54,51 @@ private static int Main(string[] args)
+            }
+        }
+
+        [SkipLocalsInit]
+        private static void ReadBytes()
+        {
+            using Stream stdin = Console.OpenStandardInput();
+            Span<byte> buffer = stackalloc byte[0x200];
+            Unsafe.InitBlock(ref MemoryMarshal.GetReference(buffer), 0, 0x200);
+            Span<char> hex = stackalloc char[] { '\0', '\0' };
```
Maybe a little intense for a test utility but it's still very readable. Definitely open to reverting to a less performance focused implementation if that's preferred.

SeeminglyScience commented on src/System.Management.Automation/engine/AsyncByteStreamDrainer.cs:
```diff
@@ -0,0 +1,76 @@
+
+    private Task? _readToBufferTask;
+
+    public AsyncByteStreamDrainer(IStreamSource streamSource, SpanAction<byte, object?> callback, object? callbackArg, Action completedCallback)
```
So I tried to work in `CopyToAsync`. It would work, *but* both `Process` instances would need to be made before I can run that. That means I'd need to call `NativeCommandProcessor.Prepare` in the opposite order for native commands (e.g. upstream *first*) which in turn makes it more challenging to revert behavior when merging stderr to stdout. I think the logic is simpler when we don't need the `Process` instance created until we actually get data to write downstream

daxian-dbw replied:
`stdOutDrainer.BeginReadChunks()` is called in `InitNativeProcess` which happens in `Prepare`. Doesn't this mean that the downstream native process's input stream needs to be already opened when the current native command processor is still running `Prepare`?

As of today, the downstream native command's `_nativeProcess` is not yet created when the `BeginReadChunks()` from the upstream native command starts to run, so there seems to be a race condition: `_callback` gets called within `BeginReadChunks()`, which calls `_nativeProcess.StandardInput.BaseStream` on the downstream native command processor while `_nativeProcess` is still `null`.

SeeminglyScience committed:
Fix for steppable pipeline

SeeminglyScience committed:
Remove the bits from TestExe I didn't end up using

SeeminglyScience committed:
Clean up over abstracted stream interface

SeeminglyScience committed:
Clean up NativeCommandProcessor.cs

- Re-arrange based on the existing regions in the file
- Add doc comments

SeeminglyScience committed:
Add more doc comments

SeeminglyScience committed:
Merge branch 'master' into preserve-native-stdout

SeeminglyScience committed:
Fix style violation

SeeminglyScience committed:
Fix new line in tests on *nix

SeeminglyScience committed:
Fix non-native merging redirection into file

SeeminglyScience committed:
Fix expected bytes in test

SeeminglyScience commented:
I don't think the remaining errors are related but I'll continue looking into them unless someone knows they can be ignored

Marking as ready for review

SteveL-MSFT reviewed (changes requested):
Initial feedback, still reviewing

SeeminglyScience committed:
Rename AsyncByteStreamDrainer

SeeminglyScience committed:
Rename BeginReadChucks

SeeminglyScience commented on src/System.Management.Automation/engine/AsyncByteStreamDrainer.cs:
```diff
@@ -0,0 +1,80 @@
+/// </summary>
+internal sealed class AsyncByteStreamDrainer : IDisposable
+{
+    private const int DefaultBufferSize = 1024;
```
I went with the default that .NET uses when you call `Process.BeginOutputReadLine`. Note that this isn't the underlying pipe's buffer size, so it's likely that it will still use the OS's default.

SeeminglyScience commented on src/System.Management.Automation/engine/AsyncByteStreamDrainer.cs:
```diff
@@ -0,0 +1,80 @@
+
+    public Task EOF => _readToBufferTask ?? Task.CompletedTask;
+
+    public void BeginReadChucks()
```
It sure was! Fixed, good catch!

SeeminglyScience commented on src/System.Management.Automation/engine/AsyncByteStreamDrainer.cs:
```diff
@@ -0,0 +1,80 @@
+                    break;
+                }
+            }
+            catch (IOException)
```
This was modeled after dotnet/runtime's `AsyncStreamReader` that is used when you call `Process.BeginOutputReadLine`. The relevant snippet is this block [here](https://github.com/dotnet/runtime/blob/6823bc2a78e3228680e9ff9808cf0e8bd4a413b2/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs#L103-L117).

I can copy the comment over if that's desired, but the answer seems to be "any platform specific edge case we don't handle"

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1656,21 +1750,33 @@ public ProcessOutputHandler(Process process, BlockingCollection<ProcessOutputObj
+            }
+
+            stdOutDrainer = _stdOutDrainer = stdOutDestination.Bind(stdOutSource);
+            stdOutDrainer.BeginReadChunks();
```
`stdOutDrainer.BeginReadChunks()` is called in `InitNativeProcess` which happens in `Prepare`. Doesn't this mean that the downstream native process's input stream needs to be already opened when the current native command processor is still running `Prepare`?

As of today, the downstream native command's `_nativeProcess` is not yet created when the `BeginReadChunks()` from the upstream native command starts to run, so there seems to be a race condition: `_callback` gets called within `BeginReadChunks()`, which calls `_nativeProcess.StandardInput.BaseStream` on the downstream native command processor while `_nativeProcess` is still `null`.

SeeminglyScience replied:
> `stdOutDrainer.BeginReadChunks()` is called in `InitNativeProcess` which happens in `Prepare`. Doesn't this mean that the downstream native process's input stream needs to be already opened when the current native command processor is still running `Prepare`?
> 
> As of today, the downstream native command's `_nativeProcess` is not yet created when the `BeginReadChunks()` from the upstream native command starts to run, so there seems to be a race condition: `_callback` gets called within `BeginReadChunks()`, which calls `_nativeProcess.StandardInput.BaseStream` on the downstream native command processor while `_nativeProcess` is still `null`.

You're absolutely correct. Initially I was forcing the upstream commands `Prepare` method to be invoked first but while cleaning up I really wanted to not change the order there.

Unfortunately a minimal hello world rust app is enough to hit a NRE here, so great catch! I'm thinking a sync primitive to block until ready will still be more clean than trying to force the upstream to be prepared first. Working on that now

SeeminglyScience replied:
Alright I've added some synchronization. I'm less convinced that it's cleaner than forcing downstream commands to `Prepare` first, let me know what you think. I'm still stress testing but it's looking good so far.

SeeminglyScience committed:
Fix behavior change when feature is turned off

SteveL-MSFT reviewed (approved):
My concerns have been addressed, just need to take care of Dongbo's feedback

SeeminglyScience committed:
Fix race condition and address feedback

SeeminglyScience commented:
> Would this CL also allow passing binary files as input to a native command? I know `Get-Content` has an `-AsByteStream` flag so it seems like `cat -AsByteStream file.bin | native.exe` should work. (Proper input redirection would be nice, I imagine there's a bug for that somewhere)

Yep!

SeeminglyScience committed:
Fix steppable pipeline

daxian-dbw commented:
@SeeminglyScience Most of my comments above are not replied. Can you please reply to my comments, so as to help me understand what changes were made? Thanks!

SeeminglyScience commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -161,7 +162,14 @@ internal static class PipelineOps
+            {
+                isNativeCommand = true;
+                nativeCommandProcessor.UpstreamIsNativeCommand = lastCommandWasNative;
+            }
```
Done!

SeeminglyScience commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1471,7 +1551,9 @@ private void CalculateIORedirection(bool isWindowsApplication, out bool redirect
-                if (IsDownstreamOutDefault(this.commandRuntime.OutputPipe))
+                if (ExperimentalFeature.IsEnabled(ExperimentalFeature.PSNativeCommandPreserveBytePipe)
+                    && StdOutDestination is null
+                    && IsDownstreamOutDefault(this.commandRuntime.OutputPipe))
```
Fixed!

SeeminglyScience committed:
Get rid of extra UpstreamIsNativeCommand check

SeeminglyScience commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1448,7 +1528,7 @@ private static bool IsDownstreamOutDefault(Pipe downstreamPipe)
         private void CalculateIORedirection(bool isWindowsApplication, out bool redirectOutput, out bool redirectError, out bool redirectInput)
         {
-            redirectInput = this.Command.MyInvocation.ExpectingInput;
+            redirectInput = this.Command.MyInvocation.ExpectingInput || UpstreamIsNativeCommand;
```
Fixed!

This was from when `NativeCommandProcessor.Prepare` was calling upstream command's `Prepare`. Now that that's fixed this check is no longer needed :tada:

SeeminglyScience committed:
Use PSObject.Base instead of is PSO

SeeminglyScience commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1909,14 +2015,34 @@ internal void Add(object input)
             {
-                AddXmlInput(input);
+                baseObjInput = pso.BaseObject;
+            }
```
Good call! Fixed ❤️

daxian-dbw commented on src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs:
```diff
@@ -890,6 +892,25 @@ private PipelineProcessor CreatePipelineProcessor()
+                    }
+
                     commandProcessorBase.RedirectShellErrorOutputPipe = this.RedirectShellErrorOutputPipe;
                     pipelineProcessor.Add(commandProcessorBase);
```
The same code is now used in `InvokePipeline`, `GetSteppablePipeline`, and `LocalPipeline.CreatePipelineProcessor()`. It would be great if it can be done in `pipelineProcessor.Add`.

SeeminglyScience replied:
Done! Much cleaner ty ❤️

daxian-dbw commented on src/System.Management.Automation/engine/AsyncByteStreamTransfer.cs:
```diff
@@ -0,0 +1,100 @@
+                break;
+            }
+
+            _callback(_buffer.Span.Slice(0, bytesRead), destinationStream);
+        }
+
+        _callbackCompleted(destinationStream);
```
Regarding `_callback` and `_callbackCompleted` are solely defined in `BytePipe.Bind`, which is not a virtual method and thus these call backs cannot be changed.
Since they cannot be changed, maybe directly call `destinationStream.Write` and `destinationStream.Close` here?

daxian-dbw replied:
@SeeminglyScience Can you please reply to this comment?

SeeminglyScience replied:
You are correct! Removed the callbacks, much cleaner now thank you!

SeeminglyScience committed:
Move native command checks to `PipelineProcessor`

SeeminglyScience committed:
Remove unnecessary variable assignment in pattern

SeeminglyScience committed:
Remove unnecessary `StdOutDestination` check

SeeminglyScience commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1471,7 +1595,8 @@ private void CalculateIORedirection(bool isWindowsApplication, out bool redirect
                 // we should not count it as a redirection.
-                if (IsDownstreamOutDefault(this.commandRuntime.OutputPipe))
+                if (StdOutDestination is null
+                    && IsDownstreamOutDefault(this.commandRuntime.OutputPipe))
```
Yep you're 100% right, fixed

SeeminglyScience commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -161,7 +162,12 @@ internal static class PipelineOps
+            if (commandProcessor is NativeCommandProcessor nativeCommandProcessor)
+            {
+                isNativeCommand = true;
+            }
```
Done!

SeeminglyScience committed:
Moved `using` statement to after the cr header

SeeminglyScience commented on src/System.Management.Automation/engine/BytePipe.cs:
```diff
@@ -0,0 +1,114 @@
+using System.Threading;
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
```
Oops! Fixed

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -31,7 +31,8 @@ internal static class PipelineOps
                                                        CommandRedirection[] redirections,
-                                                       ExecutionContext context)
+                                                       ExecutionContext context,
+                                                       bool lastCommandWasNative)
```
@SeeminglyScience I think you missed this comment. Or if this parameter is needed, please point me to the usage of it.

SeeminglyScience replied:
Fixed!

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:
```diff
@@ -161,7 +162,12 @@ internal static class PipelineOps
                                              (cmd is ScriptCommand || cmd is PSScriptCmdlet);
 
-            bool isNativeCommand = commandProcessor is NativeCommandProcessor;
+            bool isNativeCommand = false;
+            if (commandProcessor is NativeCommandProcessor)
+            {
+                isNativeCommand = true;
+            }
```
This change can be reverted
```suggestion
            bool isNativeCommand = commandProcessor is NativeCommandProcessor;
```

SeeminglyScience replied:
Done!

SeeminglyScience committed:
Removed used parameter

SeeminglyScience committed:
Revert isNativeCommand if statement expansion

SeeminglyScience committed:
Remove callbacks from AsyncByteStreamTransfer

SeeminglyScience committed:
Fix compile errors from overload change

daxian-dbw commented on src/System.Management.Automation/engine/AsyncByteStreamTransfer.cs:
```diff
@@ -0,0 +1,97 @@
+            {
+                stream = await _bytePipe.GetStream(_cts.Token);
+                destinationStream = await _destinationPipe.GetStream(_cts.Token);
+            }
+            catch (IOException)
+            {
+                return;
+            }
+            catch (OperationCanceledException)
+            {
+                return;
+            }
+
+            while (true)
+            {
+                int bytesRead;
+                try
+                {
+                    bytesRead = await stream.ReadAsync(_buffer, _cts.Token);
+                    if (bytesRead is 0)
+                    {
+                        break;
+                    }
+                }
+                catch (IOException)
+                {
+                    break;
+                }
+                catch (OperationCanceledException)
+                {
+                    break;
+                }
+
+                destinationStream.Write(_buffer.Span.Slice(0, bytesRead));
+            }
+        }
+        finally
+        {
+            destinationStream?.Close();
+        }
```
I think we can reduce to only 1 try/catch/finally block, but let me know if I did something wrong below:

```suggestion
        try
        {
            stream = await _bytePipe.GetStream(_cts.Token);
            destinationStream = await _destinationPipe.GetStream(_cts.Token);

            while (true)
            {
                int bytesRead;
                bytesRead = await stream.ReadAsync(_buffer, _cts.Token);
                if (bytesRead is 0)
                {
                    break;
                }

                destinationStream.Write(_buffer.Span.Slice(0, bytesRead));
            }
        }
        catch (IOException)
        {
            return;
        }
        catch (OperationCanceledException)
        {
            return;
        }
        finally
        {
            destinationStream?.Close();
        }
```

daxian-dbw reviewed (approved):
LGTM. Left one more suggestion, mostly ready to merge now 🚀

SeeminglyScience committed:
Merge remote-tracking branch 'upstream/master' into preserve-native-stdout

SeeminglyScience committed:
Update src/System.Management.Automation/engine/AsyncByteStreamTransfer.cs

Co-authored-by: Dongbo Wang <dongbow@microsoft.com>

SeeminglyScience committed:
Add experimental feature usage telemetry

JamesWTruher commented on src/System.Management.Automation/engine/BytePipe.cs:
```diff
@@ -103,6 +104,10 @@ internal static FileBytePipe Create(string fileName, bool append)
 
+        ApplicationInsightsTelemetry.SendExperimentalUseData(
+            ExperimentalFeature.PSNativeCommandPreserveBytePipe,
+            "native to file created");
```
This can just be `File`
reduces the payload string by a bunch and still retains information.
I suppose it could also be `OutFile`

JamesWTruher commented on src/System.Management.Automation/engine/pipeline.cs:
```diff
@@ -265,6 +267,15 @@ internal int Add(CommandProcessorBase commandProcessor)
+                        {
+                            ApplicationInsightsTelemetry.SendExperimentalUseData(
+                                ExperimentalFeature.PSNativeCommandPreserveBytePipe,
+                                "native to native created");
```
reduce to just `Pipe`?

JamesWTruher reviewed (commented):
a couple of small suggestions
being _overly_ descriptive here isn't necessarily a desired thing. It increases our payload size and ingestion. Think of the string as a simple key that can be queried for in the telemetry data (terabytes of it)
It essentially turns into a column title, but I think the string _as is_ still flows on the wire.

SeeminglyScience commented on src/System.Management.Automation/engine/pipeline.cs:
```diff
@@ -257,6 +261,33 @@ private void LogToEventLog()
+                        {
+                            ApplicationInsightsTelemetry.SendExperimentalUseData(
+                                ExperimentalFeature.PSNativeCommandPreserveBytePipe,
+                                "native to native created");
```
```suggestion
                                "p");
```

Discussed offline with Jim, we decided to go with `f` and `p`

SeeminglyScience commented on src/System.Management.Automation/engine/BytePipe.cs:
```diff
@@ -0,0 +1,115 @@
+
+        ApplicationInsightsTelemetry.SendExperimentalUseData(
+            ExperimentalFeature.PSNativeCommandPreserveBytePipe,
+            "native to file created");
```
```suggestion
            "f");
```

Discussed with Jim offline, we decided to go with `f` and `p`

SeeminglyScience committed:
Apply suggestions from code review

daxian-dbw commented:
@SeeminglyScience Thanks for pushing this change through!

hez2010 commented:
Wondering is there any chance to enable this feature by default in 7.4.0 GA eventually?

iSazonov commented:
@hez2010 Will if we get no negative feedbacks.
